Added example shell script for realtime z-shimming

jcohenadad opened the pull request:
Note: this is a work in progress, with comments about CLIs yet to be created.

jcohenadad commented on +img/read_nii.m:
```diff
@@ -0,0 +1,37 @@
+function [img,info,json] = read_nii( niiFile )
+%READ_NII Load NIfTI image and header; and, when present, the accompanying .json sidecar
```
not sure if it's of any help, but for the records here is how SPM deals with BIDS: https://en.wikibooks.org/wiki/SPM/BIDS

jcohenadad commented on Coils/Shim_Greg/ShimOpt_Greg.m:
```diff
@@ -24,11 +24,16 @@
+[ Field, Params, Specs] = ShimOpt.parseinput( varargin ) ;
+
+if isempty(Specs)
+    Shim.System.Specs = ShimSpecs_Greg() ; 
```
didn't we agree to use camelCase for variables? so having something like this:
```suggestion
    shim.System.Specs = ShimSpecs_Greg() ; 
```

or maybe that would break the code somewhere else and we should do another specific PR for changing formatting?

rtopfer replied:
baby steps!
We'll get rid of `ShimOpt_Greg.m` in a separate PR and change the case in the remaining classes.

jaystock commented:
For SMS acquisitions, would it make sense to explicitly store the slice ordering and slice group information in the .json file?  This information is in the twix meas.dat header, but I think it's also buried in the DICOM header, though it might not be parsed by normal DICOM header header scripts in Matlab.

jcohenadad commented:
> For SMS acquisitions, would it make sense to explicitly store the slice ordering and slice group information in the .json file? This information is in the twix meas.dat header, but I think it's also buried in the DICOM header, though it might not be parsed by normal DICOM header header scripts in Matlab.

Excellent idea. @rtopfer or @po09i can you look if already done by dcm2niix and if not, open an issue on their repos and document here via new issue. Thx

jcohenadad commented on +b0shim/+coils/+greg/Contents.m:
```diff
@@ -0,0 +1,8 @@
+%b0shim.coil.greg 8-channel AC/DC 3T neck coil
```
to look nice on matlab's doc:
```suggestion
%% b0shim.coil.greg
% 8-channel AC/DC 3T neck coil
```

jaystock replied:
Not sure if this is the most appropriate thread to post this, but I found a very fast way to interpolate the shim coil profile maps into the grid used by the target volume (for example the EPI volume) to prepare the coil field maps for shim optimization calculations.  I'm not sure how Ryan handled this in the original code.

`unix('flirt -in coil_field_maps.nii -ref epi_target.nii -applyxfm -usesqform -out coil_field_maps_interp')`

This seems to be ~100x faster than my previous approach which used scatteredInterpolant in Matlab.  Matlab is extremely slow if you use "scattered" coordinates for the interpolation rather than monotonically increasing points such as those generated by meshgrid.  The trick is to use the -usesqform flag in FLIRT, so that it will just apply a coordinate transform/interpolation without rotating or moving the image to register them.  

However, I'm not sure if we want to commit to relying on FSL in the GUI.   We might want to implement our own fast interpolator using the same methods as FLIRT.

jcohenadad replied:
Thank you @jaystock, I moved the discussion to a specific issue thread https://github.com/shimming-toolbox/shimming-toolbox/issues/159

jcohenadad committed:
Added example shell script for realtime z-shimming

Note: this is a work in progress, with comments about CLIs yet to be created.

jcohenadad committed:
Merge branch 'master' into jca/37-example-zshim

jcohenadad committed:
Merge branch 'master' into jca/37-example-zshim

jcohenadad commented:
> Is this script for performing realtime z-shimming, ie computing the gradient of the field map along z? I am asking because I see that the coil profiles are downloaded:

Sorry for the late reply-- indeed, we do not need the coil profile. We started drafted a general scenario, but in the case of Gz shimming (where we do not rely on a calibrated coil profile but instead trust the scanners's input Gz value), we don't need this additional coil profile.

jcohenadad committed:
Updated demo_realtime_zshimming.sh

jcohenadad committed:
Extended testing for spherical harmonics

Added example from testing data, to check resampling of SH basis on discrete grid
WIP

po09i committed:
Fixed z since it did not assign anything

po09i committed:
Added test for different values

po09i committed:
Fied fsl and more comments

po09i committed:
Added ability to process zshimming within siemens basis test

po09i commented:
I tried running the testing-data on the matlab repo to get a reference point of what the output should look like. Since only a portion of the data was uploaded to data-testing for size constraints (3 echoes and 10 time points), the matlab repo which depends on the header (which does not fit anymore) throws a lot of errors. Which acquisition did you use @evaalonsoortiz for the realtime_zshim data in this [PR](https://github.com/shimming-toolbox/data-testing/pull/4) so I can test it locally.

po09i commented:
The name of the file is ACDC112 so I'll assume this is correct

po09i committed:
Cleaned up code

po09i commented on examples/demo_realtime_zshimming.sh:
```diff
@@ -0,0 +1,64 @@
+
+# Calling FSL directly
+fsl_prepare_fieldmap SIEMENS sub-example_phasediff.nii.gz sub-example_magnitude1.nii.gz sub-example_phasediff_unwrapped.nii.gz 2.46 --nocheck
+fslmaths sub-example_phasediff_unwrapped.nii.gz -div 0.00246 sub-example_fieldmap.nii.gz
```
Should be 6.28 to go from radians/s to hz

jcohenadad committed:
Merge branch 'master' into jca/37-example-zshim

jcohenadad committed:
Fixed scaling of fieldmap

Should not be divided by 2.46ms (already done by fsl), but instead convert to Hz/s.

po09i committed:
Seperated generation of fieldmaps per timepoints to avoid fsl_prepare_fieldmap bug

po09i committed:
Merged with master

po09i committed:
updated flags for st_dicom_to_nifti now that is is available

po09i committed:
Fixed bug because of recursive copytree and added todo

jcohenadad commented on examples/demo_realtime_zshimming.sh:
```diff
@@ -0,0 +1,81 @@
+# dcm2bids -d . -o nifti -p sub-example -c ../../config/dcm2bids.json
+st_dicom_to_nifti -input . -output ../nifti -subject sub-example
+cd ../nifti/sub-example/fmap
+# TODO: Name of phase2 should be phasediff
```
open issue to address that (should be implemented in `st_dicom_to_nifti`

po09i committed:
Merge branch 'master' into jca/37-example-zshim

po09i committed:
Update with the new CLIs

po09i committed:
Merge branch 'master' into jca/37-example-zshim

po09i committed:
Add call to mask

po09i committed:
Merge branch 'master' into jca/37-example-zshim

po09i committed:
Update mask call

po09i committed:
Update shell practices

po09i committed:
Merge branch 'master' into jca/37-example-zshim

po09i committed:
Update script

po09i committed:
New script

po09i committed:
Actual new file

po09i committed:
Bootstrap fixup for images scaling from -4095 to 4095

evaalonsoortiz committed:
made series description more robust to sequence name and updated the output file in the shell script

evaalonsoortiz committed:
changed the path for the respiratory trace

po09i committed:
Handle nans so that they returned as 0

po09i committed:
Improve g sig figs and commented Gx = Gy in coordinates

po09i committed:
Update test values

po09i committed:
Quick fix of iso_to_ms where iso did not have 2 characters
